src/threads-version/big_int64.cpp

```
@@ -0,0 +1,767 @@
+    *this = b - remainder;
+  else
+    *this = remainder;
+  return *this;
+}
+
+BigInt64 BigInt64::operator%(const BigInt64 &b) const {
+  /// c=a%b;
+  BigInt64 c(*this);
+  c %= b;
+  return c;
+}
+BigInt64 &BigInt64::operator^=(const BigInt64 &b) {
+  // this^=b;
+  *this = power(*this, b);
+  return *this;
+}
+
+BigInt64 BigInt64::operator^(const BigInt64 &b) const {
+  // c=this^b
+  BigInt64 c(*this);
+  c ^= b;
+  return c;
+}
+
+BigInt64 &BigInt64::operator>>=(uint64_t n) {
+  // this>>=b;
+  rightShift(n);
+  return *this;
+}
+
+BigInt64 BigInt64::operator>>(uint64_t n) const {
+  // c=this>>b;
+  BigInt64 c(*this);
+  c >>= n;
+  return c;
+}
+BigInt64 &BigInt64::operator<<=(uint64_t n) {
+  // this<<=b;
+  leftShift(n);
+  return *this;
+}
+
+BigInt64 BigInt64::operator<<(uint64_t n) const {
+  // c=this<<b;
+  BigInt64 c(*this);
+  c <<= n;
+  return c;
+}
+
+std::ostream &operator<<(std::ostream &out, const BigInt64 &a) {
+  if (a.limbsCount() == 1) {
+    if (a.isNegative)
+      out << "-";
+    out << a.limbs[0];
+  }
+
+  else {
+    BigInt10 decimal = a.toDecimal();
+    out << decimal;
+  }
+  return out;
+}
+
+bool BigInt64::isSmallerThanUnsigned(const BigInt64 &b) const {
+  // O(n)
+  //  ans=a<b
+  if (limbsCount() > b.limbsCount())
+    // a is longer
+    return false;
+  if (limbsCount() < b.limbsCount())
+    // b is longer
+    return true;
+  // same length
+  for (int i = limbsCount() - 1; i >= 0; i--)
+    if (limbs[i] != b.limbs[i])
+      return limbs[i] < b.limbs[i];
+  // they are equal
+  return false;
+}
+
+void BigInt64::prefixPlusPlusUnsigned() {
+  //++a
+  int i, n = limbsCount();
+  // zero alll the MAX_VAL_64 lsb's
+  for (i = 0; i < n && limbs[i] == MAX_VAL_64; i++) {
+    limbs[i] = 0;
+  }
+  if (i == n)
+    // it was all MAX_VAL_64
+    limbs.push_back(1);
+  else
+    limbs[i]++;
+}
+
+void BigInt64::prefixMinusMinusUnsigned() {
+  //--a
+  if (isZero())
+    throw std::invalid_argument("cant --0");
```

Should return an error code?

---

wrapped with try catch